src/System.Net.Security/src/System/Net/CertificateValidationPal.Windows.cs

```diff
@@ -124,41 +125,28 @@ internal static X509Certificate2 GetRemoteCertificate(SafeDeleteContext security
         //
         internal static string[] GetRequestCertificateAuthorities(SafeDeleteContext securityContext)
         {
-            Interop.SspiCli.SecPkgContext_IssuerListInfoEx issuerList =
-                (Interop.SspiCli.SecPkgContext_IssuerListInfoEx)SSPIWrapper.QueryContextAttributes(
-                    GlobalSSPI.SSPISecureChannel,
-                    securityContext,
-                    Interop.SspiCli.ContextAttribute.SECPKG_ATTR_ISSUER_LIST_EX);
+            Interop.SspiCli.SecPkgContext_IssuerListInfoEx issuerList = default;
+            bool success = SSPIWrapper.QueryContextAttributes_SECPKG_ATTR_ISSUER_LIST_EX(GlobalSSPI.SSPISecureChannel, securityContext, ref issuerList, out SafeHandle sspiHandle);
 
             string[] issuers = Array.Empty<string>();
-
             try
             {
-                if (issuerList.cIssuers > 0)
+                if (success && issuerList.cIssuers > 0)
                 {
                     unsafe
                     {
-                        uint count = issuerList.cIssuers;
                         issuers = new string[issuerList.cIssuers];
-                        Interop.SspiCli.CERT_CHAIN_ELEMENT* pIL = (Interop.SspiCli.CERT_CHAIN_ELEMENT*)issuerList.aIssuers.DangerousGetHandle();
-                        for (int i = 0; i < count; ++i)
+                        var elements = new Span<Interop.SspiCli.CERT_CHAIN_ELEMENT>((void*)sspiHandle.DangerousGetHandle(), issuers.Length);
+                        for (int i = 0; i < elements.Length; ++i)
                         {
-                            Interop.SspiCli.CERT_CHAIN_ELEMENT* pIL2 = pIL + i;
-                            if (pIL2->cbSize <= 0)
+                            if (elements[i].cbSize <= 0)
                             {
-                                NetEventSource.Fail(securityContext, $"Interop.SspiCli._CERT_CHAIN_ELEMENT size is not positive: {pIL2->cbSize}");
+                                NetEventSource.Fail(securityContext, $"Interop.SspiCli._CERT_CHAIN_ELEMENT size is not positive: {elements[i].cbSize}");
                             }
-                            if (pIL2->cbSize > 0)
+                            if (elements[i].cbSize > 0)
                             {
-                                uint size = pIL2->cbSize;
-                                byte* ptr = (byte*)(pIL2->pCertContext);
-                                byte[] x = new byte[size];
-                                for (int j = 0; j < size; j++)
-                                {
-                                    x[j] = *(ptr + j);
-                                }
-
-                                X500DistinguishedName x500DistinguishedName = new X500DistinguishedName(x);
+                                byte[] x = new Span<byte>((byte*)elements[i].pCertContext, checked((int)elements[i].cbSize)).ToArray();
+                                var x500DistinguishedName = new X500DistinguishedName(x);
```

Comment:
X500DistinguishedName is allocated just to get a string out. Seems like low hanging fruit to address later? Seems like it would require a change to X500DistinguishedNameDecode in the PAL.

Reply:
X500DistinguishedName is public API in another contract. It could certainly get new API, but that's not something to be wrapped up in this PR.

Reply:
Ah, I understand now what you were suggesting, compiling in the pal code from the other library.

Reply:
Yep